lib/spack/spack/compilers/aocc.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import re
+import sys
+
+import llnl.util.lang
+
+from spack.compiler import Compiler
+
+
+class Aocc(Compiler):
+    # Subclasses use possible names of C compiler
+    cc_names = ['clang']
+
+    # Subclasses use possible names of C++ compiler
+    cxx_names = ['clang++']
+
+    # Subclasses use possible names of Fortran 77 compiler
+    f77_names = ['flang']
+
+    # Subclasses use possible names of Fortran 90 compiler
+    fc_names = ['flang']
+
+    version_argument = '--version'
+
+    @property
+    def debug_flags(self):
+        return ['-gcodeview', '-gdwarf-2', '-gdwarf-3', '-gdwarf-4',
+                '-gdwarf-5', '-gline-tables-only', '-gmodules', '-gz', '-g']
+
+    @property
+    def opt_flags(self):
+        return ['-O0', '-O1', '-O2', '-O3', '-Ofast', '-Os', '-Oz', '-Og',
+                '-O', '-O4']
+
+    @property
+    def link_paths(self):
+        link_paths = {'cc': 'aocc/clang',
+                      'cxx': 'aocc/clang++',
+                      'f77': 'aocc/flang',
+                      'fc': 'aocc/flang'}
+
+        return link_paths
+
+    @property
+    def verbose_flag(self):
+        return "-v"
+
+    @property
+    def openmp_flag(self):
+        return "-fopenmp"
+
+    @property
+    def cxx11_flag(self):
+        return "-std=c++11"
+
+    @property
+    def cxx14_flag(self):
+        return "-std=c++14"
+
+    @property
+    def cxx17_flag(self):
+        return "-std=c++17"
+
+    @property
+    def c99_flag(self):
+        return '-std=c99'
+
+    @property
+    def c11_flag(self):
+        return "-std=c11"
+
+    @property
+    def cc_pic_flag(self):
+        return "-fPIC"
+
+    @property
+    def cxx_pic_flag(self):
+        return "-fPIC"
+
+    @property
+    def f77_pic_flag(self):
+        return "-fPIC"
+
+    @property
+    def fc_pic_flag(self):
+        return "-fPIC"
+
+    required_libs = ['libclang']
+
+    @classmethod
+    @llnl.util.lang.memoized
+    def extract_version_from_output(cls, output):
```

Comment:
Other than this function, this appears to use most of the logic from `clang.py`. Like `apple_clang.py` you could have the `aocc` compiler be a subclass of `clang`, or did you try that at some point and run into problems?

Like the PR which added `apple-clang` (https://github.com/spack/spack/pull/17110), this should also update the `extract_version_from_output` method of the `clang.py` compiler class (to avoid identifying `AMD clang` as `clang`): 

```
class Clang(Compiler):
...
    def extract_version_from_output(cls, output):
        ver = 'unknown'
        if 'Apple' in output or 'AMD' in output:  <------ update this to check for 'AMD'
            return ver
```


`Aocc.extract_version_from_output` does not need the check

```
        if 'Apple' in output:
             return loc_ver # i.e. 'unknown'
```

because the regular expressions in this implementation contain `"AMD clang version"` which will not match for instances of `apple-clang`.

Reply:
Please note that 
(1)  AOCC as sub-class as clang is not helpful for me, even though all the functions look similar I must rewrite them as the compiler version of AOCC and Clang are different. Also, I cannot inherit clang's fc and f77 mappings 

(2) To distinguish between AOCC and clang , just like Apple i have added code for AMD in the file  "spack/lib/spack/spack/compilers/clang.py" and commited the same along with my changes 
--snip--
if ('Apple' in output) or ('AMD' in output):
--snip--

Also, you can see that I am avoding Apple in "lib/spack/spack/compilers/aocc.py" and clang also gets avoided as clang will not have 'AMD in it's version"

--snip--
def extract_version_from_output(cls, output):
        loc_ver = 'unknown'
        if 'Apple' in output:
            return loc_ver

        match = re.search(
            r'AMD clang version ([^ )]+)',
            output
        )
        if match:
            loc_ver = output.split('AOCC_')[1].split('-')[0]
        return loc_ver
--snip--

Finally in "./lib/spack/spack/compilers/apple_clang.py" they are checking for "^Apple" which will avoid both clang and AOCC 

--snip--
def extract_version_from_output(cls, output):
        ver = 'unknown'
        match = re.search(
            # Apple's LLVM compiler has its own versions, so suffix them.
            r'^Apple (?:LLVM|clang) version ([^ )]+)',
            output,
            # Multi-line, since 'Apple clang' may not be on the first line
            # in particular, when run as gcc, it seems to output
            # "Configured with: --prefix=..." as the first line
            re.M,
        )
        if match:
            ver = match.group(match.lastindex)
        return ver
--sinp--

Based on the above explanation, there will not be clash between AMD, Clang, and Apple_clang